File: luxtensor/test_signed_tx.js
// LuxTensor - Test Script for eth_sendRawTransaction with Signed Transactions
// Run with: node test_signed_tx.js

const crypto = require('crypto');

// Hardhat Account #0 (for testing only - NEVER use in production!)
const PRIVATE_KEY = 'ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';
const FROM_ADDRESS = 'f39Fd6e51aad88F6F4ce6aB8827279cffFb92266';
const TO_ADDRESS = '70997970C51812dc3A010C7d01b50e0d17dc79C8';

const RPC_URL = 'http://localhost:8545';

// Helper to convert hex string to bytes
function hexToBytes(hex) {
    hex = hex.replace('0x', '');
    const bytes = [];
    for (let i = 0; i < hex.length; i += 2) {
        bytes.push(parseInt(hex.substr(i, 2), 16));
    }
    return Buffer.from(bytes);
}

// Helper to pad number to fixed bytes
function padBytes(num, length) {
    const hex = num.toString(16).padStart(length * 2, '0');
    return hexToBytes(hex);
}

// Create raw transaction bytes
// Format: from(20) + nonce(8) + to(20) + value(16) + gas(8) + data_len(4) + data + v(1) + r(32) + s(32)
function createRawTransaction(from, nonce, to, value, gas, data, v, r, s) {
    const parts = [
        hexToBytes(from),           // 20 bytes
        padBytes(nonce, 8),         // 8 bytes
        hexToBytes(to),             // 20 bytes
        padBytes(value, 16),        // 16 bytes (u128)
        padBytes(gas, 8),           // 8 bytes
        padBytes(data.length / 2, 4), // 4 bytes (data length)
        hexToBytes(data || ''),     // variable
        Buffer.from([v]),           // 1 byte
        hexToBytes(r),              // 32 bytes
        hexToBytes(s),              // 32 bytes
    ];
    return Buffer.concat(parts);
}

// Simple signing (for demo - in production use proper secp256k1)
function signMessage(message, privateKey) {
    // This is a placeholder - in production, use proper ECDSA with secp256k1
    const hash = crypto.createHash('sha256').update(message).digest();
    const sig = crypto.createHash('sha256').update(Buffer.concat([hash, hexToBytes(privateKey)])).digest();
    return {
        r: sig.toString('hex').padStart(64, '0'),
        s: sig.toString('hex').padStart(64, '0'),
        v: 0
    };
}

// Create signing message (same format as Transaction::signing_message in Rust)
function createSigningMessage(nonce, from, to, value, gasPrice, gasLimit, data) {
    const parts = [
        padBytes(nonce, 8),
        hexToBytes(from),
        to ? hexToBytes(to) : Buffer.alloc(20, 0),
        padBytes(value, 16),
        padBytes(gasPrice, 8),
        padBytes(gasLimit, 8),
        hexToBytes(data || '')
    ];
    return Buffer.concat(parts);
}

async function sendRpcRequest(method, params) {
    const response = await fetch(RPC_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            jsonrpc: '2.0',
            method: method,
            params: params,
            id: 1
        })
    });
    return response.json();
}

async function main() {
    console.log('🔗 LuxTensor - Testing eth_sendRawTransaction');
    console.log('================================================\n');

    // Step 1: Get current nonce
    console.log('1. Getting current nonce...');
    const nonceResult = await sendRpcRequest('eth_getTransactionCount', ['0x' + FROM_ADDRESS, 'latest']);
    const nonce = parseInt(nonceResult.result, 16) || 0;
    console.log(`   Nonce: ${nonce}\n`);

    // Step 2: Create transaction
    console.log('2. Creating transaction...');
    const txParams = {
        nonce: nonce,
        from: FROM_ADDRESS,
        to: TO_ADDRESS,
        value: BigInt('1000000000000000000'),  // 1 ETH
        gas: 21000,
        gasPrice: 1,
        data: ''
    };
    console.log(`   From:  0x${txParams.from}`);
    console.log(`   To:    0x${txParams.to}`);
    console.log(`   Value: ${txParams.value} wei (1 ETH)\n`);

    // Step 3: Sign transaction
    console.log('3. Signing transaction...');
    const signingMessage = createSigningMessage(
        txParams.nonce,
        txParams.from,
        txParams.to,
        Number(txParams.value),
        txParams.gasPrice,
        txParams.gas,
        txParams.data
    );
    const signature = signMessage(signingMessage, PRIVATE_KEY);
    console.log(`   Signature R: 0x${signature.r.substring(0, 16)}...`);
    console.log(`   Signature S: 0x${signature.s.substring(0, 16)}...`);
    console.log(`   V: ${signature.v}\n`);

    // Step 4: Create raw transaction
    console.log('4. Creating raw transaction bytes...');
    const rawTx = createRawTransaction(
        txParams.from,
        txParams.nonce,
        txParams.to,
        Number(txParams.value),
        txParams.gas,
        txParams.data,
        signature.v,
        signature.r,
        signature.s
    );
    console.log(`   Raw TX length: ${rawTx.length} bytes`);
    console.log(`   Raw TX hex: 0x${rawTx.toString('hex').substring(0, 40)}...\n`);

    // Step 5: Send raw transaction
    console.log('5. Sending eth_sendRawTransaction...');
    const sendResult = await sendRpcRequest('eth_sendRawTransaction', ['0x' + rawTx.toString('hex')]);

    if (sendResult.error) {
        console.log(`   ❌ Error: ${sendResult.error.message}\n`);
    } else {
        console.log(`   ✅ TX Hash: ${sendResult.result}\n`);
    }

    // Step 6: Check balances
    console.log('6. Checking balances after 5 seconds...');
    await new Promise(r => setTimeout(r, 5000));

    const fromBalance = await sendRpcRequest('eth_getBalance', ['0x' + FROM_ADDRESS, 'latest']);
    const toBalance = await sendRpcRequest('eth_getBalance', ['0x' + TO_ADDRESS, 'latest']);

    console.log(`   From balance: ${fromBalance.result}`);
    console.log(`   To balance:   ${toBalance.result}\n`);

    console.log('================================================');
    console.log('✅ Test complete!');
}

main().catch(console.error);

File: luxtensor/test_signed_tx.py
"""Quick test: Send a signed EIP-155 transaction and check if it gets mined."""
import json
import time
import requests
from eth_account import Account
from eth_account.signers.local import LocalAccount

RPC_URL = "http://127.0.0.1:8545"
CHAIN_ID = 8898  # LuxTensor chain ID from config.toml

# Genesis funded account (from config.toml prefunded_accounts)
FUNDED_KEY = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
funded: LocalAccount = Account.from_key(FUNDED_KEY)

RECIPIENT = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"

def rpc(method, params=None):
    r = requests.post(RPC_URL, json={
        "jsonrpc": "2.0", "id": 1,
        "method": method, "params": params or []
    }, timeout=10)
    data = r.json()
    if "error" in data:
        raise Exception(f"RPC error: {data['error']}")
    return data.get("result")

def main():
    print(f"=== Signed TX Mining Test ===")
    print(f"Sender:    {funded.address}")
    print(f"Recipient: {RECIPIENT}")
    print(f"Chain ID:  {CHAIN_ID}")

    # Get current block number
    block_before = int(rpc("eth_blockNumber"), 16)
    print(f"\nCurrent block: {block_before}")

    # Get nonce
    nonce = int(rpc("eth_getTransactionCount", [funded.address, "latest"]), 16)
    print(f"Nonce: {nonce}")

    # Get balance
    balance = int(rpc("eth_getBalance", [funded.address, "latest"]), 16)
    print(f"Balance: {balance / 1e18:.4f} ETH")

    # Build and sign transaction
    tx = {
        "nonce": nonce,
        "to": RECIPIENT,
        "value": 1_000_000_000_000_000_000,  # 1 ETH
        "gas": 21000,
        "gasPrice": 1_000_000_000,  # 1 Gwei
        "chainId": CHAIN_ID,
    }

    signed = Account.sign_transaction(tx, FUNDED_KEY)
    raw_tx = signed.raw_transaction.hex()
    # Ensure 0x prefix
    if not raw_tx.startswith("0x"):
        raw_tx = "0x" + raw_tx

    print(f"\nSigned TX hash: {signed.hash.hex()}")
    print(f"Raw TX length:  {len(raw_tx)} chars")

    # Send raw transaction
    print("\nSending eth_sendRawTransaction...")
    tx_hash = rpc("eth_sendRawTransaction", [raw_tx])
    print(f"TX hash returned: {tx_hash}")

    # Wait for the transaction to be mined
    print("\nWaiting for TX to be mined...")
    for i in range(30):  # Wait up to 30 * 3 = 90 seconds
        time.sleep(3)
        receipt = rpc("eth_getTransactionReceipt", [tx_hash])
        block_now = int(rpc("eth_blockNumber"), 16)
        print(f"  [{i+1}] Block: {block_now}, Receipt: {'YES' if receipt else 'not yet'}")
        if receipt:
            print(f"\n✅ TX MINED in block {int(receipt['blockNumber'], 16)}!")
            print(f"   Status: {receipt.get('status', 'N/A')}")
            print(f"   Gas used: {int(receipt.get('gasUsed', '0x0'), 16)}")

            # Verify balance changed
            new_balance = int(rpc("eth_getBalance", [funded.address, "latest"]), 16)
            recipient_balance = int(rpc("eth_getBalance", [RECIPIENT, "latest"]), 16)
            print(f"\nSender balance:    {new_balance / 1e18:.4f} ETH (was {balance / 1e18:.4f})")
            print(f"Recipient balance: {recipient_balance / 1e18:.4f} ETH")
            return True

    print("\n❌ TX NOT MINED after 90 seconds!")
    return False

if __name__ == "__main__":
    success = main()
    exit(0 if success else 1)
